packages/components/src/ui/color-picker/use-deprecated-props.ts

```diff
@@ -0,0 +1,161 @@
+/**
+ * External dependencies
+ */
+import colorize, { ColorFormats } from 'tinycolor2';
+// eslint-disable-next-line no-restricted-imports
+import type { ComponentProps } from 'react';
+import memoize from 'memize';
+
+/**
+ * WordPress dependencies
+ */
+import { useCallback, useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import type ColorPicker from './component';
+
+type ColorPickerProps = ComponentProps< typeof ColorPicker >;
+
+/**
+ * @deprecated
+ */
+type LegacyColor =
+	| string
+	| {
+			hex: string;
+			hsl: ColorFormats.HSL | ColorFormats.HSLA;
+			hsv: ColorFormats.HSV | ColorFormats.HSVA;
+			rgb: ColorFormats.RGB | ColorFormats.RGBA;
+			/**
+			 * @deprecated
+			 */
+			oldHue: number;
+			/**
+			 * @deprecated
+			 */
+			source: 'hex';
+			draftHex: string;
+			draftHsl: ColorFormats.HSL | ColorFormats.HSLA;
+			draftRgb: ColorFormats.RGB | ColorFormats.RGBA;
+	  };
+
+/**
+ * @deprecated
+ */
+export interface LegacyProps {
+	color: LegacyColor;
+	/**
+	 * @deprecated
+	 */
+	onChangeComplete: ( colors: LegacyColor ) => void;
+	/**
+	 * @deprecated
+	 */
+	oldHue: string;
+	className: string;
+	/**
+	 * @deprecated
+	 */
+	disableAlpha: boolean;
+}
+
+function isLegacyProps( props: any ): props is LegacyProps {
+	return (
+		typeof props.onChangeComplete !== 'undefined' ||
+		typeof props.color === 'string' ||
+		typeof props.hex === 'string'
+	);
+}
+
+function getColorFromLegacyProps(
+	props: LegacyProps
+): ColorFormats.HSL | ColorFormats.HSLA {
+	if ( typeof props.color === 'string' ) {
+		return colorize( props.color ).toHsl();
+	}
+
+	return props.color.hsl;
+}
+
+function toHsv(
+	color: colorize.Instance
+): ColorFormats.HSV | ColorFormats.HSVA {
+	const { h, s, v, a } = color.toHsv();
+
+	return {
+		h: Math.round( h ),
+		s: Math.round( s * 100 ),
+		v: Math.round( v * 100 ),
+		a,
+	};
+}
+
+const transformHslToLegacyColor = memoize(
+	( hsla: ColorFormats.HSL | ColorFormats.HSLA ): LegacyColor => {
+		const color = colorize( hsla );
+		const rawHex = color.toHex();
+		const rgb = color.toRgb();
+		const hsv = toHsv( color );
+		const hsl = hsla;
+
+		const isTransparent = rawHex === '000000' && rgb.a === 0;
+
+		const hex = isTransparent ? 'transparent' : `#${ rawHex }`;
+
+		return {
+			hex,
```

Comment:
When comparing the return value of `transformHslToLegacyColor` against the output format of the `color` prop in the legacy `ColorPicker`, I notice that:

- the `color` prop is missing
- the `draftHex`, `draftHsl` and `draftRgb` props are extra

In fact, the shape of the value object passed to `onChangeComplete` callback is derived in the [`colorToState` utility](https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/packages/components/src/color-picker/utils.js#L59-L67) — should we mirror this function's return value here?

Reply:
Ah, good catch, I misread this code and thought the `draft*` ones were also being passed, but they're just passed to `setState`: 

https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/packages/components/src/color-picker/index.js#L150-L159

I also missed the `color` property. I'll fix both in one commit.

Reply:
The `color` prop is definitely problematic for changing to use `colord` instead of `tinycolor2` isn't it?

Reply:
30456bc

Reply:
> The color prop is definitely problematic for changing to use colord instead of tinycolor2 isn't it?

I'm afraid so. The value of the `color` property seems to have the following shape

```json
{
  "_originalInput": {
    "h": 0,
    "s": 61.326530612244895,
    "v": 75.139146567718,
    "a": 1,
    "source": "rgb"
  },
  "_r": 191.60482374768088,
  "_g": 74.1002328575215,
  "_b": 74.1002328575215,
  "_a": 1,
  "_roundA": 1,
  "_format": "hsv",
  "_ok": true,
  "_tc_id": 85
}
```

Writing a `colord` to `tinycolor2` adapter for the format of this object is definitely possible, but I'm not sure it would be a good idea, since the shape of the `tinycolor2` is not guaranteed to stay the same. What do you think?

Reply:
I think it's something we should just "break"... You're absolutely right, if `tinycolor2` changed its implementation we'd be locked into the same issue. Might as well just remove the `color` property or deprecate it and change it to the `colord` instance.

Reply:
> I think it's something we should just "break"

I agree that this may be a "good" breaking change. It feels also safe enough, since the returned object already has all of the common formats already there (`hex`, `hsl`, ...) and therefore I don't believe that many users would be using the `tinycolor2` object?

I would personally remove the `color` property altogether — it's best not to have any public APIs depending on 3rd party libraries, otherwise we're going to have the same issue when switching from `colord` to the next library.

Reply:
That's in fact a big issue on the onChangeComplete function :( That function seems to pass too many things to the caller including a tiny color object. I think we should solve that. I guess the most common use case is accessing the hex property.
So I agree I think we should deprecate the color property from onChangeComplete, not include any colord or tiny color instance, even if there we were not replacing libraries I would say we should do this change. Maybe the object can be { hex, rgb, hsl, hsv } and with that we probably handle 99.9% of the use cases. And I guess we can create a dev note to warn API users.

Reply:
Moving the conversation over to https://github.com/WordPress/gutenberg/issues/34286